### ISSUE TITLE
HTTP client POST request?

### ISSUE THREAD

**Issue by Assert1** [warmcat/libwebsockets#598]
I am using lws as it primary designed for (WS and HTTP server) in my software for 1 year, and it works fine, thank you for this nice library. But now, i need to make a simple HTTP client that creates POST requests to another server. Of course i can use another library to do this, but it will be great to do it with lws, and do not add any new dependencies to my project. So...

I am made my own lws based HTTP client that works fine with my existing lws based server using GET method. May be i missed something, but as i understand from lws debugging, there is no way to make full client side POST request with large body content at this moment. After changing method to POST, and modifying request header to add Content-Type and Content-Length in LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER, client stucks (lws_callback_on_writable is not working there, and i have no idea when and how to write POST body data) before ESTABLISHED state, waiting for some server data to finish handshake. But this data can be sended by the server only after reciving everything from the client, including POST body data.

If we trying to modify header, and add small POST request body right there, server side calls LWS_CALLBACK_HTTP_BODY and LWS_CALLBACK_HTTP_BODY_COMPLETION as expected (except crash if we returning -1 in LWS_CALLBACK_HTTP_BODY_COMPLETION to drop connection), we can create response and send it to the client, that can recive it and process. Reserved header space is enought for my requirements, so i can use this workaround for the POST requests, but... may be there is a more straight way to send POST body data to the server?

**Comment by lws-team** [comment]
Right... you can set the .method in the client connect info as you found, that way you can definitely do GET (the test client supports it) and to some extent do POST, at least as far as doing an http request with POST method.

For server side there is no problem, again the test server demo has POST and file upload function with a nice api.

But for client POST payload there are no states on lws client side to support it atm (that is why callback when writable cannot act), after forming the http request he sends it and then gets ready to receive the reply in state LWSCM_WSCL_WAITING_SERVER_REPLY: that's all that is needed for ws upgrade and GET.  It needs something inbetween like a new connection state after sending the http headers where the client connection can choose to stay until he sent everything he intended in the writable callback then indicate somehow he can move on to the existing state of waiting for the response headers.

I'll take a look at it.

**Comment by lws-team** [comment]
It's implemented here

**Comment by Assert1** [comment]
Works nice, thank you!

**Comment by lws-team** [comment]
Great.

**Comment by lws-team** [comment]
When I woke up I realized having lws do the lws_callback_on_writable(wsi); itself was a bad idea... it works well but it removes the chance for the user code to wait for something to deliver it the data it wants to send.

So I changed the patch to require lws_callback_on_writable(wsi); explicitly in LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER and LWS_CALLBACK_CLIENT_HTTP_WRITEABLE if you will send a client http body.  The test client, comments and docs are updated to reflect that as well.
